Create new admin tool

aarongable opened the pull request:
Create a new administration tool "bin/admin" as a successor to and replacement of "admin-revoker".

This new tool supports all the same fundamental capabilities as the old admin-revoker, including:
- Revoking by serial, by batch of serials, by incident table, and by private key
- Blocking a key to let bad-key-revoker take care of revocation
- Clearing email addresses from all accounts that use them

Improvements over the old admin-revoker include:
- All commands run in "dry-run" mode by default, to prevent accidental executions
- All revocation mechanisms allow setting the revocation reason, skipping blocking the key, indicating that the certificate is malformed, and controlling the number of parallel workers conducting revocation
- All revocation mechanisms do not parse the cert in question, leaving that to the RA
- Autogenerated usage information for all subcommands
- A much more modular structure to simplify adding more capabilities in the future
- Significantly simplified tests with smaller mocks

The new tool has analogues of all of admin-revokers unit tests, and all integration tests have been updated to use the new tool instead. A future PR will remove admin-revoker, once we're sure SRE has had time to update all of their playbooks.

Fixes https://github.com/letsencrypt/boulder/issues/7135
Fixes https://github.com/letsencrypt/boulder/issues/7269
Fixes https://github.com/letsencrypt/boulder/issues/7268
Fixes https://github.com/letsencrypt/boulder/issues/6927
Part of https://github.com/letsencrypt/boulder/issues/6840

~~DO NOT MERGE before https://github.com/letsencrypt/boulder/pull/7275~~

mcpherrinm commented:
admin clear-email fixes the most important part of https://github.com/letsencrypt/boulder/issues/6864 as well

mcpherrinm reviewed (commented):
I'm beginning to review, but some notes on the overall operation of the command:

1. Some sort of "admin self-check" would be useful to ensure the configuration is valid, SA/RA/DB can be contacted successfully, etc.  That could be run automatically as a post-deploy check of some sort. We wouldn't want to find out during an incident that a seldom-used operation wasn't working because it, eg, used the db connection directly and that wasn't working for some reason.

2. How should output logs work?  This is an incomplete thought, but we probably want both stdout and syslog.

3. usage string should have a 1-liner explaining what each subcommand does.  Could the usage string be automatically generated by iterating over all subcommands? That probably would mean using a map instead of a switch statement.

4. It doesn't have to happen here, but avoiding DB connections is IMO a high-priority follow-up ask for this, as that'll simplify deployment significantly

mcpherrinm commented on cmd/admin/key.go:
```diff
@@ -0,0 +1,87 @@
+		return errors.New("the -private-key flag is required")
+	}
+
+	spkiHash, err := a.spkiHashFromPrivateKey(*privKey)
```
It seems like it would be easy and possibly useful to allow specifying the SPKI hash directly, in case I have that and not the private-key (and don't have a cert to use with revoke-cert)

aarongable replied:
This is https://github.com/letsencrypt/boulder/issues/7267, not doing it as part of this PR (since I'm not adding any truly new functionality here, just unifying existing functionality).

pgporada reviewed (commented):
I've not finished my review, but here's some very basic TUI/UX finds.

mcpherrinm commented on cmd/admin/email.go:
```diff
@@ -0,0 +1,65 @@
+// accounts, it does not exit on failure, preferring to continue and make as
+// much progress as possible.
+func (a *admin) subcommandClearEmail(ctx context.Context, args []string) error {
+	subflags := flag.NewFlagSet("clear-email", flag.ExitOnError)
```
I think this should be a command like "update-email", with a -clear flag.

That provides an obvious place to add an update option in the future.

I think it would be best to have a flag with a list of registration IDs to update, and a separate lookup command to run ahead of time.  That way the actual clear is fast (since we can just touch those rows), and can code-review the implementation plan to remove a user's contact information and be sure which registrations it'll touch.

We probably want the registrations flag to be optional in case we need to remove in bulk (eg, if somebody asks to remove an email from hundreds or thousands of accounts, perhaps because they register each time)

aarongable replied:
I think these are great ideas for a follow-up change. Right now this change does not add any new functionality, and does not meaningfully change any of the existing functionality, with the exception of unifying which revocation commands can take which flags. I really want to keep this PR scoped to "replace admin-revoker", not "and add a bunch of new utilities". Can these ideas be tracked on https://github.com/letsencrypt/boulder/issues/6864 instead?

mcpherrinm replied:
I don't think I'm suggesting additional functions here; this is suggesting how to name and structure the new commands to allow for future expandability. 

I'm just saying you should name the command "update-email" with a mandatory -clear flag.

 I don't really care whether it's in this PR or a future one

mcpherrinm commented:
I'm a bit worried overall about the Dry Run handling:  It seems error-prone and easy to have a bug where you forget an early return or mess up a conditional, leading to mutations in what's supposed to be a dry run.

Perhaps there should be a safety wrapper around the admin.rac and admin.sac which checks the dry run flag and panics if a mutating RPC is called.

aarongable commented on cmd/admin/admin.go:
```diff
@@ -0,0 +1,66 @@
+package main
```
Hmm thanks for pointing these out. Honestly, my preferred solution would be to remove this subcommand entirely; it's hard for me to image it being super useful when someone could get this info from the playbook that told them to run `admin revoke-cert` in the first place.

mcpherrinm replied:
Instead of a list-reason, what if this information was in the usage for "revoke-cert"?  
Can revoke-cert -reason take a string from this set, so we're not handling the raw numbers in the CLI at all?

aarongable replied:
I've removed list-reasons and changed revoke-cert -reason to take a string (and updated the usage to list the allowed strings).

aarongable commented:
> Some sort of "admin self-check" would be useful to ensure the configuration is valid, SA/RA/DB can be contacted successfully, etc. That could be run automatically as a post-deploy check of some sort. We wouldn't want to find out during an incident that a seldom-used operation wasn't working because it, eg, used the db connection directly and that wasn't working for some reason.

We've still [failed to implement this idea elsewhere](https://github.com/letsencrypt/boulder/issues/6548) for over a year now. It's a good idea, but gRPC doesn't have good mechanisms for this. The "list-reasons" subcommand _does_ require that the config parses correctly and that the admin tool can successfully construct DB and gRPC client connections, but we don't have a boilerplate way to exercise those connections.

> How should output logs work? This is an incomplete thought, but we probably want both stdout and syslog.

They go through the same log package as all other boulder tools, so they [can be configured](https://github.com/letsencrypt/boulder/pull/7276/files#diff-5de7c6d21b96a8e6dc32593546fab462e730dd1befe78170a09d00ded8ec50aaR35-R38) to go to both stdout and syslog.

> usage string should have a 1-liner explaining what each subcommand does. Could the usage string be automatically generated by iterating over all subcommands? That probably would mean using a map instead of a switch statement.

Yeah, I had this same thought. Iterating over all subcommands means having a registry, having a registry means (ideally) having a way for subcommands to register themselves, and that means (ideally) having each subcommand in a separate package and doing registration via `init()`. I'm totally happy to go this way if folks want, I just didn't want to overengineer it on the first draft.

> It doesn't have to happen here, but avoiding DB connections is IMO a high-priority follow-up ask for this, as that'll simplify deployment significantly

Yep, definitely a concern, but also definitely a concern for a follow-up change. That should be tracked in a separate bug.

> I'm a bit worried overall about the Dry Run handling: It seems error-prone and easy to have a bug where you forget an early return or mess up a conditional, leading to mutations in what's supposed to be a dry run.
> Perhaps there should be a safety wrapper around the admin.rac and admin.sac which checks the dry run flag and panics if a mutating RPC is called.

To do this, I'd probably want to replace the dbConn by splitting the SAClient into an SAClient and an SAROClient. Then all calls to the read-write SAClient and the RAClient could be intercepted by dry-run mode. Like maybe in dry-run it wouldn't even instantiate those connections, it would just replace those objects with stubs.

mcpherrinm commented:
> I'm totally happy to go this way if folks want, I just didn't want to overengineer it on the first draft.
Totally understand, and I agree there's no need to overengineer.  A simple static map initialized in main.go is almost no more work than the current switch statement on line 101, so I think there's a pretty comfortable middle ground if you want.

pgporada commented on cmd/admin/cert.go:
```diff
@@ -0,0 +1,260 @@
+	// Flags specifying the input method for the certificates to be revoked.
+	serial := subflags.String("serial", "", "Revoke the certificate with this hex serial")
+	incidentTable := subflags.String("incident-table", "", "Revoke all certificates whose serials are in this table")
+	serialsFile := subflags.String("serials-file", "", "Revoke all certificates whose hex serials are in this file")
```
What do you mean by "hex serials"?

aarongable replied:
Serials in hexadecimal encoding

pgporada commented on cmd/admin/cert.go:
```diff
@@ -0,0 +1,260 @@
+	if err != nil {
+		return nil, fmt.Errorf("computing SPKI hash: %w", err)
+	}
+
```
Lines 172-180 are a duplicate of  `a.spkiHashFromPrivateKey` in `//cmd/admin/key.go`. These lines could become
```
spkiHash, err := a.spkiHashFromPrivateKey(privkeyFile)
if err != nil {
	return err
}
```

pgporada commented on cmd/admin/cert.go:
```diff
@@ -0,0 +1,260 @@
+
+	// General flags relevant to all certificate input methods.
+	parallelism := subflags.Uint("parallelism", 10, "Number of concurrent workers to use while revoking certs")
+	reason := subflags.Uint("reason", 0, "Revocation reason (must be 0, 1, 4, 5, or 9)")
```
```suggestion
	reason := subflags.Uint("reason", 0, "Revocation reason (must be 0, 1, 4, 5, or 9). The default is 0 (unspecified).")
```

aarongable replied:
The flag package appends "(default: 0)" automatically.

pgporada commented on cmd/admin/cert.go:
```diff
@@ -0,0 +1,260 @@
+	}
+
+	var serials []string
+	_, err = a.dbMap.Select(ctx, &serials, "SELECT certSerial FROM keyHashToSerial WHERE keyHash = ? AND certNotAfter > NOW()", spkiHash[:])
```
The clause `AND certNotAfter > NOW()` is a divergence from [admin-revoker](https://github.com/letsencrypt/boulder/blob/2f114f8a9b9612a6d57774cab40e3620e3ec4533/cmd/admin-revoker/main.go#L450). That's intended?

aarongable replied:
Yep. We do not need to revoke certificates that have already expired, and this clause makes the query handle a smaller result set.

pgporada commented on cmd/admin/key.go:
```diff
@@ -0,0 +1,87 @@
+}
+
+func (a *admin) blockSPKIHash(ctx context.Context, spkiHash []byte, comment string) error {
+	var count int
```
Should there be a [query](https://github.com/letsencrypt/boulder/blob/2f114f8a9b9612a6d57774cab40e3620e3ec4533/cmd/admin-revoker/main.go#L409-L420) to determine if the SPKI is already in the blockedKeys table and avoid doing the work below?

aarongable replied:
Done!

aarongable commented on test/v2_integration.py:
```diff
@@ -1590,9 +1590,12 @@ def test_admin_revoker_cert():
+        "-reason", "1"])
 
     # Wait for OCSP response to indicate revocation took place
     verify_ocsp(cert_file.name, "/hierarchy/intermediate-cert-rsa-a.pem", "http://localhost:4002", "revoked")
```
Added the reason code to the `verify_ocsp()` calls.

aarongable commented:
I think I've addressed all comments so far. Thanks so much @pgporada and @mcpherrinm! I think the new registration and extra safety around dry-run is really good.

pgporada commented on cmd/admin/cert_test.go:
```diff
@@ -0,0 +1,282 @@
+}
+
+// SerialsForIncident returns a fake gRPC stream client object which itself
+// will return the mockSAWithIncident's serials in order.
```
This "in order" I assume means "returns mockSAWithIncident's serials in the order in which they were received from the gRPC stream" ?

aarongable replied:
It means "return's mockSAWithIncident's serials in the order in which they appear in its `incidentSerials` slice"

aarongable commented on cmd/admin/cert.go:
```diff
@@ -0,0 +1,244 @@
+		"-serial":         *serial != "",
+		"-incident-table": *incidentTable != "",
+		"-serials-file":   *serialsFile != "",
+		"-private-key":    *privKey != "",
```
Gonna leave this for now for the sake of closing the review loop; it can get cleaned up as part of https://github.com/letsencrypt/boulder/issues/7267
